### PR TITLE
[Design] First cut of Razor Project System (based on CPS)

### PR DESCRIPTION
**Pull request by rynowak**
@NTaylorMullen - take a look at this bad boy with regard to the design of project snapshot manager, and lets discuss any feedback that you have. This is the direction I think we should go. I want to move forward with this after making sure we can achieve something similar in VS4Mac for VS2TheMaxxxx

### CODE REVIEW

File: src/Microsoft.CodeAnalysis.Razor.Workspaces/Microsoft.CodeAnalysis.Razor.Workspaces.csproj [hunk 1]
```diff
@@ -7,6 +7,7 @@
   </PropertyGroup>
 
   <ItemGroup>
+    <ProjectReference Include="..\Microsoft.AspNetCore.Mvc.Razor.Extensions\Microsoft.AspNetCore.Mvc.Razor.Extensions.csproj" />
```
Comment (rynowak) [inline src/Microsoft.CodeAnalysis.Razor.Workspaces/Microsoft.CodeAnalysis.Razor.Workspaces.csproj]: A lot of the stuff here is a total hack job so I could experiment on you know what

File: src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/DefaultDocumentSnapshot.cs [hunk 2]
```diff
@@ -0,0 +1,21 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+namespace Microsoft.CodeAnalysis.Razor.ProjectSystem
+{
+    internal class DefaultDocumentSnapshot : DocumentSnapshot
```
Comment (rynowak) [inline src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/DefaultDocumentSnapshot.cs]: The idea is that we'd eventually implement enough stuff here to build a `RazorProjectFileSystem` based on what the project system knows about.

File: src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/DefaultProjectSnapshotUpdateListener.cs [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+using System;
+using System.IO;
+using Microsoft.AspNetCore.Mvc.Razor.Extensions;
+using Microsoft.AspNetCore.Razor.Language;
+using Microsoft.CodeAnalysis.CSharp;
+
+namespace Microsoft.CodeAnalysis.Razor.ProjectSystem
+{
+    internal class DefaultProjectSnapshotUpdateListener : ProjectSnapshotUpdateListener
```
Comment (rynowak) [inline src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/DefaultProjectSnapshotUpdateListener.cs]: Ignore everything related to this. It's 100%haxxxxx

File: test/testapps/RazorIdeTesting/Controllers/HomeController.cs [hunk 4]
```diff
@@ -0,0 +1,43 @@
+﻿using System;
```
Comment (rynowak) [inline test/testapps/RazorIdeTesting/Controllers/HomeController.cs]: Ignore everything below here, I checked in a project that I created from the template.

**Comment by NTaylorMullen** [comment]
Lets do an in-person design discussion Monday. This is a "holy shit" amount of code 😉 I want to make sure I understand it all so we can ensure this works for VS mac.

**Comment by rynowak** [comment]
Closing this out in favor of some smaller PRs.

**Pull request closed without merging**